packages/core/src/lib/health_manager.ts

```diff
@@ -0,0 +1,74 @@
+import {
+  HealthStatus,
+  type IHealthManager,
+  NodeHealth,
+  type ProtocolHealth,
+  Protocols
+} from "@waku/interfaces";
+
+class HealthManager implements IHealthManager {
+  public static instance: HealthManager;
+  private readonly health: NodeHealth;
+
+  private constructor() {
+    this.health = {
+      overallStatus: HealthStatus.Unhealthy,
+      protocolStatuses: new Map()
+    };
+  }
+
+  public static getInstance(): HealthManager {
+    if (!HealthManager.instance) {
+      HealthManager.instance = new HealthManager();
+    }
+    return HealthManager.instance;
+  }
+
+  public getHealthStatus(): HealthStatus {
+    return this.health.overallStatus;
+  }
+
+  public getProtocolStatus(protocol: Protocols): ProtocolHealth | undefined {
+    return this.health.protocolStatuses.get(protocol);
+  }
+
+  public updateProtocolHealth(
+    protocol: Protocols,
+    connectedPeers: number
+  ): void {
+    let status: HealthStatus = HealthStatus.Unhealthy;
+    if (connectedPeers == 1) {
+      status = HealthStatus.MinimallyHealthy;
+    } else if (connectedPeers >= 2) {
+      status = HealthStatus.SufficientlyHealthy;
+    }
+
+    this.health.protocolStatuses.set(protocol, {
+      name: protocol,
+      status: status,
+      lastUpdate: new Date()
+    });
+
+    this.updateOverallHealth();
+  }
+
+  private updateOverallHealth(): void {
+    const relevantProtocols = [Protocols.LightPush, Protocols.Filter];
+    const statuses = relevantProtocols.map(
+      (p) => this.getProtocolStatus(p)?.status
+    );
+
+    if (statuses.some((status) => status === HealthStatus.Unhealthy)) {
+      this.health.overallStatus = HealthStatus.Unhealthy;
+    } else if (
+      statuses.some((status) => status === HealthStatus.MinimallyHealthy)
+    ) {
+      this.health.overallStatus = HealthStatus.MinimallyHealthy;
+    } else {
+      this.health.overallStatus = HealthStatus.SufficientlyHealthy;
+    }
+  }
+}
+
+export const getHealthManager = (): HealthManager =>
```

Comment:
Can't we just use `HealthManager.getInstance()` in the code?

Reply:
but then `export const getHealthManager = (): HealthManager =>` is not the fix as it is prone to the same race conditions because it doesn't have any lock inside. 

I would use one function, but I am not insisting